### PR TITLE
DOC: Fix description of dtype default in linspace

### LINKED ISSUE #8597: linspace dtype documentation misleading or a bug
The linspace documentation for the dtype argument says:

    dtype : dtype, optional
        The type of the output array.  If `dtype` is not given, infer the data
        type from the other input arguments.

However in the code linspace:

    start = asanyarray(start) * 1.0
    stop  = asanyarray(stop)  * 1.0
    dt = result_type(start, stop, float(num))
    if dtype is None:
        dtype = dt

Which means the dtype is only inferred from the other arguments if they are a floating-point type (and then not always if they are too small). It will never infer an integral or other type and the float(num) causes it to be at least float64 on more systems.

I think the correct solution here is to just state the it defaults to a floating-point type in the documentation. The reason that is probably important instead of actually inferring the dtype is that if someone has:

    np.linspace(0, 1, 100)

the inferred type would be int which is clearly not what was expected here. The only other reasonable solution would be calculate `step` first and determine if it is an integral step along with the start and stop being integral as well. Something along the following lines (which isn't complete, just an example):

    start = asanyarray(start)
    stop  = asanyarray(stop)
    if dtype is None and start.dtype.kind in 'iu' and stop.dtype.kind in 'iu' and (start - stop) % num == 0:
        dtype = result_type(start, stop, num)
    else:
        start = start * 1.0
        stop  = stop * 1.0
        dt = result_type(start, stop, float(num))
        if dtype is None:
            dtype = dt

### PR DESCRIPTION
**Pull request by bjnath**
Clarify linspace docstring to say the default dtype will never be smaller than float.

Fixes gh-8597

### CODE REVIEW

**Review by rossbar — COMMENTED** [review]
I don't think "smaller than float" is an accurate wording - from the discussion in #8597, the main takeaway is that the inferred type will be a floating point type  in the case where the inputs (`start`, `stop`, and `step`) are all integer types (this makes sense given what linspace does, but the wording for the dtype parameter description is not very clear on this point). 

Whatever wording is chosen should also accommodate complex `start`/`stop` values as well.

**Comment by bjnath** [comment]
I think the new wording says what you believe it needs to. "Infers" is from the original and seems like the right approach;  readers can understand this means the dtype chosen will handle complex arguments correctly. But "infers" also suggests it would choose an integer for integer start and stop, and "smaller than float" indicates it won't.

**Comment by rossbar** [comment]
> "Infers" is from the original and seems like the right approach;

Agreed! No problems at all with this wording.

> But "infers" also suggests it would choose an integer for integer start and stop,

Agreed here as well - this is what I understand to be the core of #8597

> "smaller than float" indicates it won't.

I wasn't very clear in my original comment, let me try to rephrase: I think the specific wording "smaller than float" is imprecise and I had trouble interpreting it; i.e. in what sense is a floating point type smaller than an integer type? It's not clear to me how "smaller" maps onto the discussion about types. It could be that I'm just missing something.

Re: my comment about complex numbers, I was mostly referring to this comment from #8597:

> Don't think its wrong, but we can add "a floating point type inferred" or so.

I was just thinking that we should be careful about wording like this as well, because it might introduce unintended ambiguities with respect to complex types. As I understand the [NumPy type hierarchy](https://numpy.org/devdocs/reference/arrays.scalars.html), complex types are distinct from floating point types, so saying "a floating point type is inferred" would technically be incorrect in the case that `start`/`stop` are complex.

**Comment by bjnath** [comment]
Ah, we're closer than I'd thought. I see your point about "smaller," too.  I was thinking "smaller" in the sense of the promotion hierarchy. But I failed to make it clear, so I'll revisit.

>  we should be careful about wording like this as well

My thought too, same reason.

**Comment by bjnath** [comment]
2db208c rewords to 

> If `dtype` is not given, the data type is inferred from `start`, `stop`, and `num`. The inferred dtype will never be an integer; `float` is chosen even if all three arguments are  integers.

File: numpy/core/function_base.py [hunk 1]
```diff
@@ -52,8 +52,10 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
         If True, return (`samples`, `step`), where `step` is the spacing
         between samples.
     dtype : dtype, optional
-        The type of the output array.  If `dtype` is not given, infer the data
-        type from the other input arguments.
+        The type of the output array.  If `dtype` is not given, the data type
+        is inferred from `start`, `stop`, and `num`. The inferred dtype will
```
Comment (rossbar) [inline numpy/core/function_base.py:56]: I would leave `num` off the list - it must be an integer, but for the purposes of determining the dtype it is cast to float internally.

**Review by rossbar — COMMENTED** [review]
This is an improvement over the last iteration and essentially addresses #8597. I'm not sure a change in the documentation is necessarily an improvement as I think the original wording (combined with an understanding of what linspace does) should be sufficient to convince users that not to expect an integer type from linspace by default. 

It would be nice to be able to say "infer the appropriate inexact type from the data", but I don't think that `np.inexact` is necessarily well-known enough to clarify things for most users. 

In conclusion - Let's see what other reviewers think! If the consensus is that it's an improvement to explicitly say that `linspace` won't return integer arrays by default, then the wording here accomplishes that.

**Comment by Qiyu8** [comment]
I found three more "infer the data type from the other input arguments" statement which should clarified as well.

- line 1324 from `_add_newdocs.py`
- line 205 and line 300 from `function_base.py`

File: numpy/core/function_base.py (hunk 1 above)
Comment (bjnath) [reply numpy/core/function_base.py:56]: Agreed, `num` is dont-care.

**Comment by bjnath** [comment]
> not to expect an integer type from linspace
@rossbar, why would I not expect th dtype here to have been integer?

> In [2]: np.linspace(start = 0, stop = 100, num = 5)
> Out[2]: array([  0.,  25.,  50.,  75., 100.])
>

**Comment by bjnath** [comment]
Commit 5487847 resolves https://github.com/numpy/numpy/pull/16433#discussion_r434295779.
Text is now

>     dtype : dtype, optional
>         The type of the output array.  If `dtype` is not given, the data type
>         is inferred from `start` and `stop`. The inferred dtype will
>         never be an integer; `float` is chosen even if the arguments would
>         produce an array of integers.
> 

@Qiyu8, thanks for finding these:
> I found three more "infer the data type from the other input arguments" statement which should clarified as well.

I'll revise the PR to include them once new wording is chosen.

**Comment by rossbar** [comment]
> @rossbar, why would I not expect the dtype here to have been integer?

Due to the fact that linspace evenly subdivides an interval. The fact that np.inexact types are returned is consistent with Python itself (since Py3k anyways) in that true division is the default, regardless of input types.

Thanks for the `grep` @Qiyu8 - the fact that there are more instances makes me more inclined to keep the current wording. Let's see if any other reviewers can weigh in!

**Comment by bjnath** [comment]
> Due to the fact that linspace evenly subdivides an interval. The fact that np.inexact types are returned is consistent with Python itself (since Py3k anyways) in that true division is the default, regardless of input types.

I think you've made my case. It took 35 words to explain why the behavior is obvious. 

Therefore, to many users, it *won't* be obvious. Would it kill us to spell it out? 

> the fact that there are more instances makes me more inclined to keep the current wording.

What's the reasoning here?

**Comment by bjnath** [comment]
Also, linspace is one of the first NumPy commands shown to beginners. 

However obvious the float outcome is to an experienced user, it's not going to be obvious to them.

How much testimony do we need that the default behavior confuses people? It confused me, it confused whoever raised gh-8597, and it confused [this unfortunate blogger](https://www.sharpsightlabs.com/blog/numpy-linspace/), who thought the result of
```
np.linspace(start = 0, stop = 100, num = 5)
```
was so obviously integer that they didn't bother running  it first to check:
image

Why so much hesitation over this?  What's at stake here that I'm missing?

**Comment by rgommers** [comment]
The addition of 
```
The inferred dtype will never be an integer; `float` is chosen even if the
arguments would produce an array of integers.
```
seems clear to me. Happy to go with that.

Agree with "inexact" not being useful here to clarify, too obscure.

**Comment by bjnath** [comment]
Thanks, @rossbar . How about those other instances that @Qiyu8 found -- do they still need to be looked at?

**Comment by rossbar** [comment]
They certainly can be if you think a wording update would improve clarity!
